app/controllers.js

```diff
@@ -610,6 +611,70 @@ Controller.prototype.handleApiEditProject = function(req, res) {
     });
 };
 
+// Create or overwrite an entry in phase history
+Controller.prototype.handleApiUpdatePhaseHistory = function(req, res) {
+    projectCache.getById(req.params.projectId, function(error, project) {
+        if(!project) {
+            log.debug('Error updating project. Project with given ID was not found.');
+            apiUtils.handleResultSet(res, 422,
+                new ApiResponse(null, ['Error updating project. Project with given ID was not found.'])
+            );
+        } else {
+            project.setPhaseHistoryEntry(req.body.phase, req.body.label, req.body.month, req.body.year);
+
+            projectDao.addProject(project, function(err, editedProject) {
+                log.trace('Updating project history: ' + JSON.stringify(project));
+                if(err) {
+                    log.debug('Error updating history of a project with id = ' + project.id);
+
+                    apiUtils.handleResultSet(res, 500,
+                        new ApiResponse(null, ['Error updating project history'])
+                    );
+                } else {
+                    log.debug('History of a project with id = ' + editedProject.id + ' has been updated');
+
+                    projectCache.refreshProject(editedProject.id);
+                    apiUtils.handleResultSet(res, 200,
+                        new ApiResponse({projectId: editedProject.id}, ['The project history has been updated'])
+                    );
+                }
+            });
+        }
+    });
+};
+
+// Delete an entry from phase history
+Controller.prototype.handleApiDeletePhaseHistory = function(req, res) {
+    projectCache.getById(req.params.projectId, function(error, project) {
+        if(!project) {
+            log.debug('Error updating project. Project with given ID was not found.');
+            apiUtils.handleResultSet(res, 422,
+                new ApiResponse(null, ['Error updating project. Project with given ID was not found.'])
+            );
+        } else {
+            project.removeFromPhaseHistory(req.body.phase, req.body.label);
+
+            projectDao.addProject(project, function(err, editedProject) {
+                log.trace('Updating project history: ' + JSON.stringify(project));
```

Comment:
Like before, lets not expose all of the data. This (and other log entries in this method) could also say that we are deleting a history entry, so we don't confuse this with the update method when looking through the logs